GVRf/Framework/framework/src/main/jni/engine/renderer/batch.cpp

```diff
@@ -0,0 +1,197 @@
+}
+
+Batch::Batch(int no_vertices, int no_indices) :
+        draw_count(0), vertex_count_(0), index_count_(0), vertex_limit_(no_vertices),
+        indices_limit_(no_indices), renderdata(nullptr),
+        index_offset_(0), not_batched_(false) {
+
+    vertices_.clear();
+    normals_.clear();
+    tex_coords_.clear();
+    indices_.clear();
+    matrices_.clear();
+    matrix_indices.clear();
+    render_data_set_.clear();
+
+    vertices_.reserve(no_vertices);
+    indices_.reserve(no_indices);
+    normals_.reserve(no_vertices);
+    tex_coords_.reserve(no_vertices);
+    matrix_indices.reserve(no_vertices);
+    glm::mat4 identity;
+    mesh = new Mesh();
+    glm::vec3 init_vec3(0, 0, 0);
+    glm::vec2 init_vec2(0, 0);
+    for (int i = 0; i < no_vertices; i++) {
+        vertices_.push_back(init_vec3);
+        tex_coords_.push_back(init_vec2);
+        matrix_indices.push_back(0);
+        normals_.push_back(init_vec3);
+        indices_.push_back(0);
+    }
+    for (int i = 0; i < BATCH_SIZE; i++)
+        matrices_.push_back(identity);
+}
+
+Batch::~Batch() {
+    vertices_.clear();
+    normals_.clear();
+    tex_coords_.clear();
+    indices_.clear();
+    matrices_.clear();
+    matrix_indices.clear();
+    delete renderdata;
+    delete mesh;
+}
+
+/*
+ * Add renderdata of scene object into mesh, add vertices, texcoords, normals, model matrices
+ */
+bool Batch::add(RenderData *render_data) {
+    material_ = render_data->pass(0)->material();
+    Mesh *render_mesh = render_data->mesh();
+    const std::vector<unsigned short>& indices = render_mesh->indices();
+
+    // if it is not texture shader, dont add into batch, render in normal way
+    if (material_->shader_type() != Material::ShaderType::TEXTURE_SHADER) {
+        render_data_set_.insert(render_data);
+        render_mesh->setMeshModified(false); // mark mesh clean
+        return true;
+    }
+
+    // if mesh is large, render in normal way
+    if (indices.size() + index_count_ > indices_limit_) {
+        if (draw_count > 0) {
+            return false;
+        } else {
+            render_data_set_.insert(render_data);
+            not_batched_ = true;
+            render_mesh->setMeshModified(false); // mark mesh clean
+            return true;
+        }
+    }
+    // Copy all renderData properties
+    if (draw_count == 0) {
+        if (!renderdata) {
+            renderdata = new RenderData(*render_data);
+            renderdata->set_batching(true);
+            mesh->set_vertices(vertices_);
+            mesh->set_normals(normals_);
+            mesh->set_tex_coords(tex_coords_);
+            mesh->set_indices(indices_);
+
+            // if render data has light, custom attribute will have location:3 if it does not have light,
+            // it will have location 2
+            if (renderdata->light_enabled() && renderdata->light()->enabled()) {
+                mesh->setVertexAttribLocF(3, "a_matrix_index");
```

Comment:
This might be driver dependent but normally attribute locations are sequential unless overridden by location or explicit binding to location.

As a side note, matching out/in variables between stages ((eg) out of Vertex shader to in of Fragment shader) is trickier as there are both implicit and explicit rules.